### PR TITLE
Proof of Concept: Swap out internal storage of conda-meta data

### PR DESCRIPTION
**Pull request by tswicegood**
## Tasks
* [ ] Unit tests
* [ ] Functional tests around environment-specific commands (needs to happen in `master` and be merged into this branch to ensure we're keeping a minimum level of functionality)
* [ ] Investigate possible issues with `shutil.rmtree` in Windows
* [ ] Fix `conda list` output
* [ ] Revisit `conda-meta/environment.json` and instead possibly include that data in some `{{ dist }}/env/*` directory data
* [ ] Write migration command to update an existing environment with the new data
* [ ] Make work in < Python 3

### CODE REVIEW

**Comment by tswicegood** [comment]
@ilanschnell @asmeurer here's the code I worked on last night on the plane.  It works for me™ for these commands:

* `conda create`
* `conda install`
* `conda remove`
* `conda env remove`

File: conda/install.py [hunk 1]
```diff
@@ -287,9 +288,25 @@ def create_meta(prefix, dist, info_dir, extra_info):
     meta_dir = join(prefix, 'conda-meta')
     if not isdir(meta_dir):
         os.makedirs(meta_dir)
-    with open(join(meta_dir, dist + '.json'), 'w') as fo:
-        json.dump(meta, fo, indent=2, sort_keys=True)
-
+    shutil.copytree(info_dir, join(meta_dir, dist))
+
+    if 'files' in extra_info:
+        with open(join(meta_dir, dist, 'files'), 'wb') as fp:
```
Comment (tswicegood) [inline conda/install.py]: This should probably go in an env specific directory so it doesn't overwrite anything that's already there.

File: conda/share.py [hunk 2]
```diff
@@ -108,4 +108,5 @@ def old_clone_bundle(path, prefix):
     index = get_index()
     plan.execute_actions(actions, index, verbose=False)
 
-    os.unlink(join(prefix, 'conda-meta', dist + '.json'))
+    # FIXME this might not work in weird edge cases on Windows
+    shutil.rmtree(join(prefix, 'conda-meta', dist))
```
Comment (asmeurer) [inline conda/share.py]: Just use `rm_rf`

**Comment by ilanschnell** [comment]
As I understood the discussion today, we would not write a migration tool (because one might not be able to rewrite the metadata, because of permission issues), but instead support both metadata formats for reading, and only write the new one.  Please correct me if I am wrong.

**Comment by tswicegood** [comment]
> As I understood the discussion today, we would not write a migration tool (because one might not be able to rewrite the metadata, because of permission issues), but instead support both metadata formats for reading, and only write the new one. Please correct me if I am wrong.

@ilanschnell we'll still need it, its just that we probably won't try to automatically migrate.  It will be a command that's called on an environment to "rebuild" (not sure if that's the right word, but I like the idea of `conda rebuild --name foobar`) to the new setup.  I need this for the new Anaconda Launcher so it only has to detect that it's an old environment and prompt the user to upgrade before using it.

File: conda/share.py (hunk 2 above)
Comment (tswicegood) [inline conda/share.py]: Gotcha!

**Comment by ilanschnell** [comment]
It just occurred to me that we don't even have to store the icon in the metadata at all.  Why not simply have the icon file as part of the regular package files.  For example: `share/launcher-icons/ipython-notebook.png`

This way conda doesn't have to deal with these files at all.

**Comment by ilanschnell** [comment]
It is not only `conda list` which is looking at the current data, the other commands (install, update, remove) do as well.  We have to support reading the old format, because rewriting the format will not be possible in many cases.

**Comment by tswicegood** [comment]
@ilanschnell this change takes into account `install`, `update`, and `remove`.

**Comment by ilanschnell** [comment]
When trying to create a new environment (using Python 2.7), I get the following error:
```
$ conda create -n test python=2
Fetching package metadata: ....
Solving package specifications: .
Package plan for installation in environment /Users/ilan/python/envs/test:

The following NEW packages will be INSTALLED:

    openssl:  1.0.1k-0 
    python:   2.7.9-1  
    readline: 6.2-2    
    sqlite:   3.8.4.1-0
    tk:       8.5.15-0 
    zlib:     1.2.8-0  

Proceed ([y]/n)? 

Linking packages ...
An unexpected error has occurred, please consider sending the             |  16%
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/Users/ilan/python/bin/conda", line 5, in <module>
    sys.exit(main())
  File "/Users/ilan/conda/conda/cli/main.py", line 202, in main
    args_func(args, p)
  File "/Users/ilan/conda/conda/cli/main.py", line 207, in args_func
    args.func(args, p)
  File "/Users/ilan/conda/conda/cli/main_create.py", line 50, in execute
    install.install(args, parser, 'create')
  File "/Users/ilan/conda/conda/cli/install.py", line 412, in install
    plan.execute_actions(actions, index, verbose=not args.quiet)
  File "/Users/ilan/conda/conda/plan.py", line 498, in execute_actions
    inst.execute_instructions(plan, index, verbose)
  File "/Users/ilan/conda/conda/instructions.py", line 140, in execute_instructions
    cmd(state, arg)
  File "/Users/ilan/conda/conda/instructions.py", line 84, in LINK_CMD
    link(state['prefix'], arg, index=state['index'])
  File "/Users/ilan/conda/conda/instructions.py", line 80, in link
    install.link(pkgs_dir, prefix, dist, lt, index=index)
  File "/Users/ilan/conda/conda/install.py", line 603, in link
    create_meta(prefix, dist, info_dir, meta_dict)
  File "/Users/ilan/conda/conda/install.py", line 297, in create_meta
    fp.write(bytes(f, encoding='utf-8') + b'\n')
TypeError: str() takes at most 1 argument (2 given)
```

**Comment by tswicegood** [comment]
@ilanschnell sorry -- I thought I mentioned that this only works in Python 3 right now.  It still requires that compatibility layer for writing `bytes` in py3 or `str`s in py2.

**Comment by tswicegood** [comment]
Added that as a task to make sure it doesn't get lost.

@ilanschnell I'm open to ideas on how to get that `fp.write` to work in both Python 2 and Python 3 without using `bytes()`.  My current solution would be to bring in a `compat.b()` type function that simply returns the string if the `bytes` function raises a `TypeError`, but I'm not sure if there's a better way to handle it that doesn't require that compatibility function.

**Comment by ilanschnell** [comment]
How about:
```
    fp.write('%s\n' % f)
```

**Comment by tswicegood** [comment]
@ilanschnell nope -- tried something similar, though not that exact syntax, originally.  When I use your code this is what I get:

```
Traceback (most recent call last):
  File "/Users/travis/anaconda/bin/conda", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/Users/travis/Dropbox/work/github/conda/conda/bin/conda", line 5, in <module>
    sys.exit(main())
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/cli/main.py", line 202, in main
    args_func(args, p)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/cli/main.py", line 207, in args_func
    args.func(args, p)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/cli/main_create.py", line 50, in execute
    install.install(args, parser, 'create')
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/cli/install.py", line 412, in install
    plan.execute_actions(actions, index, verbose=not args.quiet)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/plan.py", line 502, in execute_actions
    inst.execute_instructions(plan, index, verbose)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/instructions.py", line 140, in execute_instructions
    cmd(state, arg)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/instructions.py", line 84, in LINK_CMD
    link(state['prefix'], arg, index=state['index'])
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/instructions.py", line 80, in link
    install.link(pkgs_dir, prefix, dist, lt, index=index)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/install.py", line 603, in link
    create_meta(prefix, dist, info_dir, meta_dict)
  File "/Users/travis/Dropbox/work/github/conda/conda/conda/install.py", line 297, in create_meta
    fp.write('%s\n' % f)
TypeError: 'str' does not support the buffer interface
```

**Comment by tswicegood** [comment]
Actually, it looks like [this solution][1] might be the fix.  Part of the way there -- just need to make sure to call `.encode` on the string.

[1]: http://stackoverflow.com/a/5471273

**Comment by tswicegood** [comment]
@ilanschnell just pushed f9e2b31f49b452514a64f9deb3b58209a0a15f70 which should work in Python 2, though I haven't tested it.

**Comment by ilanschnell** [comment]
Ah, now I see why `fp.write('%s\n' % f)` didn't work.  Because the file was opened in `wb` mode.  When you use `w` it should work.

**Comment by ilanschnell** [comment]
So, I think have fixed all the unicode issues in my latest commit.  Tested the code on Python 2.7 and 3.4.

**Comment by tswicegood** [comment]
@ilanschnell will that work with a package that has unicode in the name?  Not that it's a good idea to do that, but I know people do.  Might also cause issues with packages that have unicode in the file names (I'm thinking of the test packages of Erlang in particular where I've run into issues with Python 2.7 and that).

**Comment by ilanschnell** [comment]
Good point.  We currently are not testing any of the conda infrastructure for package names or conda package filenames with unicode characters.  Does binstar.org support unicode package names?  How about PyPI?  I've never seen a package with a unicode name.  I'm not sure if supporting unicode package names is a good idea, I'd say no.  I think we should being this up in the next Anaconda meeting.

**Pull request closed without merging**